prover-service/src/handlers.rs

```diff
@@ -107,11 +111,98 @@ pub async fn prove_handler(
     .await
 }
 
+pub async fn prove(
+    state: &ProverServiceState,
+    witness_file: NamedTempFile,
+    public_inputs_hash: PoseidonHash,
+) -> Result<Groth16Proof, ErrorWithCode> {
+    let _span = logging::new_span("GenerateProofWithRetry");
+    let prover_unlocked = state.full_prover.lock().await;
+    let witness_file_path = witness_file.path_str().log_err()?;
+    let (proof_json, internal_metrics) = prover_unlocked
+        .prove(witness_file_path)
+        .map_err(error::handle_prover_lib_error)
+        .log_err()?;
+    metrics::GROTH16_TIME_SECS.observe((f64::from(internal_metrics.prover_time)) / 1000.0);
+
+    let rapidsnark_proof = serde_json::from_str(proof_json).map_err(anyhow::Error::from)?;
+    let proof = encode_proof(&rapidsnark_proof)?;
+
+    let g16vk = {
+        let _span = logging::new_span("PrepareVK");
+        prepared_vk(
+            &state
+                .prover_service_config
+                .test_verification_key_file_path(),
+        )
+    };
+
+    proof.verify_proof(
+        ark_bn254::Fr::from_le_bytes_mod_order(&public_inputs_hash),
+        &g16vk,
+    )?;
+
+    Ok(proof)
+}
+
 /// On all unrecognized routes, return 404.
 pub async fn fallback_handler() -> (StatusCode, &'static str) {
     (StatusCode::NOT_FOUND, "Invalid route")
 }
 
+#[allow(non_snake_case)]
+#[derive(Serialize, Deserialize, Debug)]
+struct RawVK {
```

Comment:
it should have been called "VK from snarkjs" or something, just to be more explicit?